lib/api/CDataSummarizationJsonWriter.cc

```diff
@@ -0,0 +1,170 @@
+                           m_Map.size());
+    }
+
+    TStrSizePrVecCItr begin() const { return m_Map.begin(); }
+    TStrSizePrVecCItr end() const { return m_Map.end(); }
+
+private:
+    CTrainedModel::CFeatureNameProvider m_FeatureNameProvider;
+    TStrSizePrVec m_Map;
+};
+}
+
+CDataSummarizationJsonWriter::CDataSummarizationJsonWriter(const core::CDataFrame& frame,
+                                                           core::CPackedBitVector rowMask,
+                                                           std::size_t numberColumns,
+                                                           const maths::CDataFrameCategoryEncoder& encodings)
+    : m_RowMask{std::move(rowMask)}, m_NumberColumns{numberColumns}, m_Frame{frame}, m_Encodings{encodings} {
+}
+
+void CDataSummarizationJsonWriter::addCompressedToJsonStream(TRapidJsonWriter& writer) const {
+    this->CSerializableToCompressedChunkedJson::addCompressedToJsonStream(
+        JSON_COMPRESSED_DATA_SUMMARIZATION_TAG, JSON_DATA_SUMMARIZATION_TAG, writer);
+}
+
+void CDataSummarizationJsonWriter::addToJsonStream(TGenericLineWriter& writer) const {
+
+    // Note that the data frame has extra columns added to it when running training.
+    // These are at the end and should not be serialized since it is wasteful and they
+    // are reinitialised anyway. m_NumberColumns is the number of supplied columns,
+    // i.e. the feature values and target variable.
+
+    writer.StartObject();
+
+    writer.Key(JSON_NUM_COLUMNS_TAG);
+    writer.Uint64(m_NumberColumns);
+
+    writer.Key(JSON_COLUMN_NAMES_TAG);
+    writer.StartArray();
+    for (std::size_t i = 0; i < m_NumberColumns; ++i) {
+        writer.String(m_Frame.columnNames()[i]);
+    }
+    writer.EndArray();
+
+    writer.Key(JSON_COLUMN_IS_CATEGORICAL_TAG);
+    writer.StartArray();
+    for (std::size_t i = 0; i < m_NumberColumns; ++i) {
+        writer.Bool(m_Frame.columnIsCategorical()[i]);
+    }
+    writer.EndArray();
+
+    CEncoderNameIndexMapBuilder encodingIndices{m_Frame.columnNames(),
+                                                m_Frame.categoricalColumnValues()};
+    m_Encodings.accept(encodingIndices);
+    writer.Key(JSON_ENCODING_NAME_INDEX_MAP_TAG);
+    writer.StartArray();
+    for (const auto& index : encodingIndices) {
+        writer.StartObject();
+        writer.Key(JSON_ENCODING_NAME_INDEX_MAP_KEY_TAG);
+        writer.String(index.first);
+        writer.Key(JSON_ENCODING_NAME_INDEX_MAP_VALUE_TAG);
+        writer.Uint64(index.second);
+        writer.EndObject();
+    }
+    writer.EndArray();
+
+    rapidjson::Document doc;
+    std::stringstream encodings;
+    {
+        core::CJsonStatePersistInserter inserter{encodings};
+        m_Encodings.acceptPersistInserter(inserter);
+    }
+    rapidjson::ParseResult ok(doc.Parse(encodings.str()));
+    if (static_cast<bool>(ok) == false) {
+        LOG_ERROR(<< "Failed parsing encoding json. Please report this error.");
```

Comment:
I know you've just moved this from a different file rather than writing it from scratch in this PR, but it will be frustrating if we ever see this message with no details.

```suggestion
        LOG_ERROR(<< "Failed parsing encoding json. Please report this error: "
                  << rapidjson::GetParseError_En(doc.GetParseError())));
```

Reply:
Done in [e13e88c](https://github.com/elastic/ml-cpp/pull/1907/commits/e13e88c0ffd540257f2b4cecd8ede25e1e25112c).